dataset/api.go

```diff
@@ -0,0 +1,161 @@
+package dataset
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/ONSdigital/dp-import-api/models"
+	"github.com/ONSdigital/go-ns/log"
+)
+
+var maxRetries = 5
+
+// DatasetAPI aggreagates a client and URL and other common data for accessing the API
+type DatasetAPI struct {
+	Client     *http.Client
+	URL        string
+	MaxRetries int
+	AuthToken  string
+}
+
+// New creates an DatasetAPI object
+func NewDatasetAPI(client *http.Client, datasetAPIURL, datasetAPIAuthToken string) *DatasetAPI {
+	return &DatasetAPI{
+		Client:     client,
+		URL:        datasetAPIURL,
+		MaxRetries: maxRetries,
+		AuthToken:  datasetAPIAuthToken,
+	}
+}
+
+func (api *DatasetAPI) GetURL() string {
+	return api.URL
+}
+
+// CreateInstance tells the Dataset API to create a Dataset instance
+func (api *DatasetAPI) CreateInstance(jobID, jobURL string) (instance *models.Instance, err error) {
+	path := api.URL + "/instances"
+	logData := log.Data{"URL": path}
+
+	var jsonUpload []byte
+	if jsonUpload, err = json.Marshal(models.CreateInstance(jobID, jobURL)); err != nil {
+		return
+	}
+	logData["jsonUpload"] = jsonUpload
+	jsonResult, httpCode, err := api.post(path, 0, jsonUpload)
+	logData["httpCode"] = httpCode
+	logData["jsonResult"] = jsonResult
+	if err == nil && httpCode != http.StatusOK && httpCode != http.StatusCreated {
+		err = errors.New("Bad response while creating instance")
+	}
+	if err != nil {
+		log.ErrorC("CreateInstance", err, logData)
+		return
+	}
+	instance = &models.Instance{}
+	if err = json.Unmarshal(jsonResult, instance); err != nil {
+		log.ErrorC("CreateInstance unmarshal", err, logData)
+		return
+	}
+	return
+}
+
+// UpdateState tells the Dataset API that the state of a Dataset instance has changed
+func (api *DatasetAPI) UpdateInstanceState(instanceID string, newState string) error {
+	path := api.URL + "/instances/" + instanceID
+	logData := log.Data{"URL": path}
+	jsonUpload := []byte(`{"state":"` + newState + `"}`)
```

Comment:
Should we have a state struct and use json.Marshal?

Reply:
yes, that was a bit raw, will change that, now